routing/probability_estimator_test.go

```diff
@@ -0,0 +1,162 @@
+			expectedProb, toNode, p)
+	}
+}
+
+// TestProbabilityEstimatorNoResults tests the probability estimation when no
+// results are available.
+func TestProbabilityEstimatorNoResults(t *testing.T) {
+	ctx := newEstimatorTestContext(t)
+
+	ctx.assertPairProbability(testTime, 0, 0, 0.6)
+}
+
+// TestProbabilityEstimatorOneSuccess tests the probability estimation for nodes
+// that have a single success result.
+func TestProbabilityEstimatorOneSuccess(t *testing.T) {
+	ctx := newEstimatorTestContext(t)
+
+	ctx.results = []timedPairResult{
+		{
+			timestamp: testTime.Add(-time.Hour),
+			pairResult: pairResult{
+				success: true,
+			},
+		},
+	}
+
+	// Because of the previous success, this channel keep reporting a high
+	// probability.
+	ctx.assertPairProbability(
+		testTime, 0, 100, aprioriPrevSucProb,
+	)
+
+	// Untried channels are also influenced by the success. With a
+	// aprioriWeight of 0.75, the a priori probability is assigned weight 3.
+	expectedP := (3*aprioriHopProb + 1*aprioriPrevSucProb) / 4
+	ctx.assertPairProbability(testTime, untriedNode, 100, expectedP)
+}
+
+// TestProbabilityEstimatorOneFailure tests the probability estimation for nodes
+// that have a single failure.
+func TestProbabilityEstimatorOneFailure(t *testing.T) {
+	ctx := newEstimatorTestContext(t)
+
+	ctx.results = []timedPairResult{
+		{
+			timestamp:  testTime.Add(-time.Hour),
+			pairResult: pairResult{},
+		},
+	}
+
+	// For an untried node, we expected the node probability. The weight for
+	// the failure after one hour is 0.5. This makes the node probability
+	// 0.51:
+	expectedNodeProb := (3*aprioriHopProb + 0.5*0) / 3.5
+	ctx.assertPairProbability(testTime, untriedNode, 100, expectedNodeProb)
+
+	// The pair probability decays back to the node probability. With the
+	// weight at 0.5, we expected a pair probability of 0.5 * 0.51 = 0.25.
+	ctx.assertPairProbability(testTime, 0, 100, expectedNodeProb/2)
+}
+
+// TestProbabilityEstimatorMix tests the probability estimation for nodes for
+// which a mix of successes and failures is recorded.
+func TestProbabilityEstimatorMix(t *testing.T) {
+	ctx := newEstimatorTestContext(t)
+
+	ctx.results = []timedPairResult{
+		{
+			timestamp: testTime.Add(-time.Hour),
+			pairResult: pairResult{
+				success: true,
+			},
+		},
+		{
+			timestamp:  testTime.Add(-2 * time.Hour),
+			pairResult: pairResult{},
+		},
+		{
+			timestamp:  testTime.Add(-3 * time.Hour),
+			pairResult: pairResult{},
+		},
+	}
+
+	// We expect the probability for a previously successful channel to
+	// remain high.
+	ctx.assertPairProbability(testTime, 0, 100, prevSuccessProbability)
+
+	// For an untried node, we expected the node probability to be returned.
+	// This is a weighted average of the results above and the a priori
+	// probability: 0.62.
+	expectedNodeProb := (3*aprioriHopProb + 1*prevSuccessProbability) /
+		(3 + 1 + 0.25 + 0.125)
+
+	ctx.assertPairProbability(testTime, untriedNode, 100, expectedNodeProb)
```

Comment:
it is a bit confusing that `untriedNode` and node `1` get different results here, since neither are referenced. I think the index in `ctx.result` is used to denote the `toNode` of the result. Could we instead set this explicitly (by making `results ` a map), and extract `node0 = 0`, `node1 = 1` into constants for readability?

Reply:
Yes it is a bit obscure. Followed your suggestion